compiler-core/src/javascript/pattern.rs

```diff
@@ -617,43 +704,53 @@ impl<'module_ctx, 'expression_gen, 'a> Generator<'module_ctx, 'expression_gen, '
                     if segment.type_ == crate::type_::int()
                         || segment.type_ == crate::type_::float()
                     {
-                        let details = Self::sized_bit_array_segment_details(segment)?;
+                        let details = self.sized_bit_array_segment_details(segment)?;
 
                         match segment.value.as_ref() {
                             Pattern::Int { int_value, .. }
-                                if details.size <= SAFE_INT_SEGMENT_MAX_SIZE
-                                    && details.size % 8 == 0
-                                    && offset.bits % 8 == 0 =>
+                                if details
+                                    .size
+                                    .is_literal_and(|size| size <= SAFE_INT_SEGMENT_MAX_SIZE)
+                                    && details.size.is_literal_and(|size| size % 8 == 0)
+                                    && offset.bits.is_whole_number_of_bytes() =>
                             {
+                                let size = match details.size {
+                                    BitArraySize::Literal(size) => size,
+                                    BitArraySize::Variable(_) => {
+                                        unreachable!("We already checked that it's a literal")
+                                    }
```

Comment:
We try to avoid avoid this non-functional pattern of checking that something is true and then asserting it afterwards. In this case it is being check 3 times, really it should only be done once.